File: 백준/Gold/9251. LCS/LCS.py
import sys 
input = sys.stdin.readline

a = list(input().rstrip())
b = list(input().rstrip())

dp = [[0 for _ in range(len(a)+1)]]
for i in range(len(b)):
    new_dp = [0]
    for j in range(0, len(a)):
        if b[i] == a[j]:
            new_dp.append(dp[i][j] + 1)
        else:
            new_dp.append(max(dp[i][j+1], new_dp[-1]))
    dp.append(new_dp)

print(dp[len(b)][len(a)])

File: 백준/Gold/9251. LCS/LCS.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
public class Main {
	public static void main(String[] args) throws IOException {
		// Input Start
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		String a = br.readLine();
		String b = br.readLine();
		
		// Input End
		br.close();
		
		// Solve
		int aLen = a.length();
		int bLen = b.length();
		
		int[][] dp = new int[aLen + 1][bLen + 1];
		
		// LCS DP
		for (int i = 1; i <= aLen; i++) {
			for (int j = 1; j <= bLen; j++) {
				dp[i][j] = Math.max(dp[i-1][j], dp[i][j-1]);
				
				if (a.charAt(i-1) == b.charAt(j-1)) {
					dp[i][j] = Math.max(dp[i][j], dp[i-1][j-1] + 1);
				}
			}
		}
		
		int result = dp[aLen][bLen];
		
		// Output Start
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		bw.write(result + "\n");
		bw.flush();
		
		// Output End
		bw.close();
	}
}